In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os
os.getcwd()

'/content'

In [ ]:
import numpy as np
import networkx as nx

In [ ]:
Directed_g = nx.read_edgelist(r'/content/drive/MyDrive/email-Eu-core.txt', create_using=nx.DiGraph)

PageRank Score for each node

In [ ]:
# Calculate the PageRank score for each node
pagerank_scores = nx.pagerank(Directed_g)
pagerank_scores

{'0': 0.0012754775504594832,
 '1': 0.009411560186382712,
 '2': 0.002095369381767161,
 '3': 0.0017234388700051611,
 '4': 0.002436232111121507,
 '5': 0.004525470848399022,
 '6': 0.0029192068825005272,
 '7': 0.0019464642315471185,
 '8': 0.0012364305542142788,
 '9': 0.001212215032480912,
 '10': 0.0013292723987216259,
 '11': 0.002289046520387193,
 '12': 0.0016365498076585522,
 '13': 0.002174677080911033,
 '14': 0.0017008199595549105,
 '15': 0.001589802529811306,
 '16': 0.0020210102759722526,
 '17': 0.001907055205270155,
 '18': 0.0013262819820475404,
 '19': 0.0021196946823522694,
 '20': 0.0022199973212970563,
 '21': 0.003767829276367166,
 '22': 0.0005601507079445894,
 '23': 0.002110969398523054,
 '24': 0.0010190577722610724,
 '25': 0.0012056517925602638,
 '26': 0.0010660051848652102,
 '27': 0.0017106176174544606,
 '28': 0.0029687137894348604,
 '29': 0.0015567182149975102,
 '30': 0.0017811043751992657,
 '31': 0.0011800767587410066,
 '32': 0.0010864630541162767,
 '33': 0.0008224977626537843,
 

In [ ]:
# Sort the pages based on their PageRank score from high to low
sorted_pages = sorted(pagerank_scores.items(), key=lambda x: x[1], reverse=True)

# Print the PageRank score for each node
for node, score in sorted_pages:
  print(f"Node {node} has a PageRank score of {score:.4f}")

Node 1 has a PageRank score of 0.0094
Node 130 has a PageRank score of 0.0069
Node 160 has a PageRank score of 0.0068
Node 62 has a PageRank score of 0.0053
Node 86 has a PageRank score of 0.0051
Node 107 has a PageRank score of 0.0050
Node 365 has a PageRank score of 0.0048
Node 121 has a PageRank score of 0.0047
Node 5 has a PageRank score of 0.0045
Node 129 has a PageRank score of 0.0045
Node 183 has a PageRank score of 0.0043
Node 64 has a PageRank score of 0.0042
Node 434 has a PageRank score of 0.0042
Node 532 has a PageRank score of 0.0041
Node 128 has a PageRank score of 0.0041
Node 106 has a PageRank score of 0.0040
Node 21 has a PageRank score of 0.0038
Node 166 has a PageRank score of 0.0037
Node 301 has a PageRank score of 0.0036
Node 82 has a PageRank score of 0.0035
Node 227 has a PageRank score of 0.0035
Node 333 has a PageRank score of 0.0035
Node 211 has a PageRank score of 0.0035
Node 377 has a PageRank score of 0.0034
Node 105 has a PageRank score of 0.0034
Node 256 

Authority and Hub score for each node

In [ ]:
# Compute the HITS algorithm for the graph
hub_scores, authority_scores= nx.hits(Directed_g, max_iter= 100, normalized = True)

Hub score for each node

In [ ]:
# Sort the hub score high to low
sorted_hubs = dict(sorted(hub_scores.items(), key=lambda item: item[1], reverse=True))
# Print the hub score for each node
for node, score in sorted_hubs.items():
    print(f"Node {node} has a Hub score of {score:.6f}")

Node 160 has a Hub score of 0.010629
Node 82 has a Hub score of 0.009617
Node 121 has a Hub score of 0.009530
Node 107 has a Hub score of 0.008788
Node 62 has a Hub score of 0.008233
Node 249 has a Hub score of 0.008018
Node 434 has a Hub score of 0.007541
Node 183 has a Hub score of 0.007200
Node 86 has a Hub score of 0.007003
Node 114 has a Hub score of 0.006398
Node 105 has a Hub score of 0.006320
Node 211 has a Hub score of 0.006317
Node 129 has a Hub score of 0.006263
Node 21 has a Hub score of 0.006185
Node 87 has a Hub score of 0.006105
Node 142 has a Hub score of 0.005976
Node 283 has a Hub score of 0.005945
Node 333 has a Hub score of 0.005935
Node 212 has a Hub score of 0.005905
Node 83 has a Hub score of 0.005796
Node 282 has a Hub score of 0.005794
Node 405 has a Hub score of 0.005693
Node 169 has a Hub score of 0.005562
Node 115 has a Hub score of 0.005468
Node 63 has a Hub score of 0.005423
Node 58 has a Hub score of 0.005384
Node 424 has a Hub score of 0.005380
Node 166 

Authority score for each node

In [ ]:
# Sort the authority score high to low
sorted_authority = dict(sorted(authority_scores.items(), key=lambda item: item[1], reverse=True))
# Print the authority score for each node
for node, score in sorted_authority.items():
    print(f"Node {node} has a authority score of {score:.6f}")

Node 160 has a authority score of 0.007220
Node 107 has a authority score of 0.006898
Node 62 has a authority score of 0.006696
Node 434 has a authority score of 0.006485
Node 121 has a authority score of 0.006472
Node 183 has a authority score of 0.006041
Node 128 has a authority score of 0.005948
Node 249 has a authority score of 0.005729
Node 256 has a authority score of 0.005704
Node 129 has a authority score of 0.005678
Node 283 has a authority score of 0.005624
Node 82 has a authority score of 0.005478
Node 106 has a authority score of 0.005459
Node 114 has a authority score of 0.005443
Node 87 has a authority score of 0.005425
Node 166 has a authority score of 0.005395
Node 211 has a authority score of 0.005305
Node 105 has a authority score of 0.005188
Node 212 has a authority score of 0.005182
Node 115 has a authority score of 0.005129
Node 169 has a authority score of 0.004991
Node 280 has a authority score of 0.004844
Node 81 has a authority score of 0.004800
Node 86 has a a

Comparing the results obtained from both the algorithms in parts 1 and 2 based on the node
scores.

From the results above, the pagerank and authorities work upon incoming nodes
but still shows different nodes as important because pagerank takes nodes with
maximum in-degree i.e. for node 1 whereas, authority score is calculated
using maximum incoming nodes from the hubs, so that doesn’t includes total
incoming nodes.
In Hubs, the Outgoing nodes are referred and thus, 160 has maximum
OutDegree nodes
Other than that we can observe that node 160 has miximum hub score and authority score which shows that it is a highly connected and influential node in the network. This suggests that it is both central and authoritative, and plays a key role in the network. 